### DISCUSSION TITLE
i also write code properly and some times  i am copying form this code but always shows the AttributeError: 'numpy.ndarray' object has no attribute 'to'

### DISCUSSION THREAD

**Discussion by vivek152oo3** [mrdbourke/pytorch-deep-learning#568] [Q&A]
/# Make predictions with the model
untrained_preds = model_0(X_test.to(device))
print(f"Length of predictions: {len(untrained_preds)}, Shape: {untrained_preds.shape}")
print(f"Length of test samples: {len(y_test)}, Shape: {y_test.shape}")
print(f"\nFirst 10 predictions:\n{untrained_preds[:10]}")
print(f"\nFirst 10 test labels:\n{y_test[:10]}")
   
output
AttributeError                            Traceback (most recent call last)
[<ipython-input-260-1c061636fa33>](https://localhost:8080/#) in <cell line: 2>()
      1 # Make predictions with the model
----> 2 untrained_preds = model_0(X_test.to(device))
      3 print(f"Length of predictions: {len(untrained_preds)}, Shape: {untrained_preds.shape}")
      4 print(f"Length of test samples: {len(y_test)}, Shape: {y_test.shape}")
      5 print(f"\nFirst 10 predictions:\n{untrained_preds[:10]}")

AttributeError: 'numpy.ndarray' object has no attribute 'to'

 i am dont understand what type error 
![Uploading Screenshot 2023-07-25 164515.png…]()

**Comment by Perian-Yan** [comment]
Hi,
I think X_test is ndarray istead of tensor here. You can run `X_test.dtype` to see its type. 
To convert from numpy.ndarray to tensor, you can use [`torch.from_numpy()`](https://pytorch.org/docs/stable/generated/torch.from_numpy.html).
Hope this helps :smile:

**Reply by prudhv16** [reply]
I guess we can use below code for device agnostic code
```
with torch.inference_mode():
  untrained_preds  = torch.tensor(X_test) if device == "cpu" else model_0(X_test.to(device))
```

**Comment by Haji-Suleman** [comment]
Your error means:

X_test is a NumPy array
.to(device) only works on PyTorch tensors

Fix:

Just convert it to tensor first:

import torch

```
X_test = torch.tensor(X_test, dtype=torch.float32)
untrained_preds = model_0(X_test.to(device))

```
That’s it
